README.md
# IBM-isam-infomap README

The IBM-isam-infomap extension can be installed on Visual studio code and is used to manage infomap scripts to increase development speed. 

## Features

Features include:
- Uploading mapping rule content
- Deploy changes
- Reload federation/aac runtime
- Clear trace log file
- download trace log file


## Requirements

Visual studio code 1.4.1

## Extension Usage

Commands are triggered using CONTROL + SHIFT + p

First, the ISAM Primary master hostname and credentials must be set to the extension.

- ISAM register host:
Register hostname and credentials in the extension for the amount of tim Visual Studio Code is running
- ISAM Register mapping rule:
Register existing mapping rule to be used in later commands
- ISAM Deploy changes:
Deploy changes
- ISAM upload mapping rule:
upload the current focussed text editor content to the registered mapping rule
- ISAM Reload federation runtime:
Reload the federation runtime
- ISAM Clear trace file:
Clear the trace.log file
- ISAM Download trace file:
Opens a new temporary document and downloads the current trace file
- ISAM upload, deploy reload:
upload mapping rule, deploy and reload in one command
## Installation
1. Place the folder in \Microsoft VS Code\resources\app\extensions


CHANGELOG.md
# Change Log

All notable changes to the "ibm-isam-infomap" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

src/extension.ts
// Author: Sander meyfroot
// email: sander.meyfroot@is4u.be

import * as vscode from 'vscode';
import * as request from "request-promise-native";
	let mappingruleId = 0;
	let mappingruleName = "";
	var username = "";
	var password = "";
	let applianceHost = "";
	let auth = "Basic " + new Buffer(username + ":" + password).toString("base64");
	
export async function downloadTrace(){
	vscode.window.showInformationMessage('Downloading trace file');
	(async () => {
		const baseUrl = 'https://' + applianceHost + '/isam/application_logs/federation/runtime/trace.log';
		var options = {
			method: "GET",
			uri: baseUrl,
			insecure: true,
			"rejectUnauthorized": false, 
			headers : {
				"Authorization" : auth,
				"Accept" : "application/json",
				"user-agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/74.0.3729.169 Safari/537.36"
			}
		};
		const result = await request.get(options);
		await vscode.commands.executeCommand('workbench.action.files.newUntitledFile');
		let editor = vscode.window.activeTextEditor;
		if (editor){
			editor.edit(editBuilder => {
				if (editor){
					editBuilder.insert(new vscode.Position(0,0),JSON.parse(result).contents);
				}
			});	
		}
		vscode.window.showInformationMessage('Trace file downloaded');
		
	  })();
}
export async function clearTrace(){
	vscode.window.showInformationMessage('Clearing trace file');
	(async () => {
		const baseUrl = 'https://' + applianceHost + '/isam/application_logs/federation/runtime/trace.log?action=clear';
		
		var options = {
			method: "DELETE",
			uri: baseUrl,
			insecure: true,
			"rejectUnauthorized": false, 
			headers : {
				"Authorization" : auth,
				"Accept" : "application/json",
				"user-agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/74.0.3729.169 Safari/537.36"
			}
		};
		const result = await request.delete(options);
		vscode.window.showInformationMessage('Trace file cleared');
	  })();
	
}

export async function deployChanges(isChain: boolean | undefined){
	vscode.window.showInformationMessage('Deploying changes');
		(async () => {
			const baseUrl = 'https://' + applianceHost + '/isam/pending_changes';
			
			var options = {
				method: "PUT",
				uri: baseUrl,
				insecure: true,
				"rejectUnauthorized": false, 
				headers : {
					"Authorization" : auth,
					"Accept" : "application/json",
					"user-agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/74.0.3729.169 Safari/537.36"
				}
			};
			const result = await request.put(options);
			vscode.window.showInformationMessage('Changes deployed');
			if(isChain){
				await reloadfedRuntime(true);
			}
		  })();
}
export async function uploadMappingRule(isChain: boolean | undefined){
	if(mappingruleId !==  undefined && mappingruleId !== 0){
		vscode.window.showInformationMessage('Got mapping rule identifier, uploading...');
	
		let editor = vscode.window.activeTextEditor;
		let text = "";
		if(editor !== undefined){
			text = editor.document.getText();
		}
		console.log(text);
		if (text !== undefined){
			text = text.replace(/\"/g,'\\"');	
		}
		
		(async () => {
			const baseUrl = 'https://' + applianceHost + '/iam/access/v8/mapping-rules/' + mappingruleId;
			console.log(baseUrl);
			let body_to_send = "{\"content\":\"" + text + "\"}";
			console.log(body_to_send);
			var options = {
				body: body_to_send,
				method: "PUT",
				uri: baseUrl,
				insecure: true,
				"rejectUnauthorized": false, 
				headers : {
					"Authorization" : auth,
					"content-type":"application/json",
					"Accept" : "application/json",
					"user-agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/74.0.3729.169 Safari/537.36"
				}
			};
			const result = await request.put(options);
			vscode.window.showInformationMessage('Upload done');
			if(isChain){
				await deployChanges(true);
			
			}
		})();
	}else{
		vscode.window.showInformationMessage('No mapping rule in cache, register mapping rule');
	}

}
export async function reloadfedRuntime(isChain: boolean | undefined){
	vscode.window.showInformationMessage('Reloading federation runtime');
	(async () => {
		const baseUrl = 'https://' + applianceHost + '/mga/runtime_profile/local/v1';
		var options = {
			body: "{operation: \"reload\"}",
			method: "PUT",
			uri: baseUrl,
			insecure: true,
			"rejectUnauthorized": false, 
			headers : {
				"Authorization" : auth,
				"Accept" : "application/json",
				"user-agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/74.0.3729.169 Safari/537.36"
			}
		};
		const result = await request.put(options);
		vscode.window.showInformationMessage('Reloaded federation runtime');
		if(isChain){
			vscode.window.showInformationMessage('Action chain completed');
		}
})();
}
export async function getMappingRuleId(){
		(async () => {
			const baseUrl = 'https://' + applianceHost + '/iam/access/v8/mapping-rules/?filter=name%20equals%20' + mappingruleName;
			var options = {
				method: "GET",
				uri: baseUrl,
				insecure: true,
				"rejectUnauthorized": false, 
				headers : {
					"Authorization" : auth,
					"Accept" : "application/json",
					"user-agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/74.0.3729.169 Safari/537.36"
				}
			};
			const result = await request.get(options);
			let mappingid = JSON.parse(result)[0].id;
			if(mappingid !== undefined){
				if (mappingid !== 0){
					mappingruleId = mappingid;
					vscode.window.showInformationMessage('saved mappingrule ID');
				}else{
					vscode.window.showInformationMessage('Mapping rule not found');

					mappingruleId = 0;
				}
			}
			console.log(mappingid);
	})();
}

export async function showInputBox() {
	const result = await vscode.window.showInputBox({
		value: mappingruleName
	});
	vscode.window.showInformationMessage(`Got: ${result}`);
	if (result === undefined){
		mappingruleName = "";
	}else{
		mappingruleName = result;
		getMappingRuleId();
	}
}
export async function registerHost() {
	vscode.window.showInformationMessage('Setting hostname');
	var result;
	if (applianceHost === undefined || applianceHost === ""){
		result = await vscode.window.showInputBox({
			value: "<set your appliance hostname>"
		});
	}else{
		result = await vscode.window.showInputBox({
			value: applianceHost
		});
	} 
	vscode.window.showInformationMessage(`Got hostname: ${result}`);

	if (result !== undefined && result !== ""){
		applianceHost = result;

		if(username === undefined || username === ""){
			vscode.window.showInformationMessage('Setting username');
			result = await vscode.window.showInputBox({
				value: "admin"
			});
		}else{
			vscode.window.showInformationMessage('Setting username');
			result = await vscode.window.showInputBox({
				value: username
			});
		}
		vscode.window.showInformationMessage(`Got username: ${result}`);
		if (result !== undefined && result !== ""){
			username = result;
			
			if (password === undefined || password ===""){
				vscode.window.showInformationMessage('Setting password');
				result = await vscode.window.showInputBox({
					value: "<set your password>"
				});
			}else{
				result = await vscode.window.showInputBox({
					value: password
				});
			}
			if (result !== undefined && result !== ""){
				password = result;
			}
		}
	}
}
export function settingsValid(){
	if(username !== undefined && username !== "" && password !== undefined && password !== "" && applianceHost !== undefined && applianceHost !== ""){
		return true;
	}else{
		vscode.window.showInformationMessage('No credentials/host found, execute ISAM Register host command');
		return false;
	}
}
export function activate(context: vscode.ExtensionContext) {

	let disposable = vscode.commands.registerCommand('extension.uploadMappingrule',()=>{
		if (settingsValid()){
			uploadMappingRule(false);
		}
	});
	disposable = vscode.commands.registerCommand('extension.registerMappingrule',()=>{
		if (settingsValid()){
			showInputBox();
		}
	});
	disposable = vscode.commands.registerCommand('extension.reloadRuntime', () => {
		if(settingsValid()){
			reloadfedRuntime(false);
		}
	});
	disposable = vscode.commands.registerCommand('extension.deployChanges', () =>{
		if(settingsValid()){
			deployChanges(false);
		}
	});
	disposable = vscode.commands.registerCommand('extension.doAll',async () => {
		if (settingsValid()){
			await uploadMappingRule(true);
			
		}

	});
	disposable = vscode.commands.registerCommand('extension.downloadTrace',() => {
		if(settingsValid()){
			downloadTrace();
		}
	});
	disposable = vscode.commands.registerCommand('extension.clearTrace',() => {
		if (settingsValid()){
			clearTrace();
		}
	});
	disposable = vscode.commands.registerCommand('extension.registerHost',() =>{
		registerHost();
	});

	context.subscriptions.push(disposable);
}

// this method is called when your extension is deactivated
export function deactivate() {}


out/extension.js
"use strict";
// Author: Sander meyfroot
// email: sander.meyfroot@is4u.be
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const vscode = require("vscode");
const request = require("request-promise-native");
let mappingruleId = 0;
let mappingruleName = "";
var username = "";
var password = "";
let applianceHost = "";
let auth = "Basic " + new Buffer(username + ":" + password).toString("base64");
function downloadTrace() {
    return __awaiter(this, void 0, void 0, function* () {
        vscode.window.showInformationMessage('Downloading trace file');
        (() => __awaiter(this, void 0, void 0, function* () {
            const baseUrl = 'https://' + applianceHost + '/isam/application_logs/federation/runtime/trace.log';
            var options = {
                method: "GET",
                uri: baseUrl,
                insecure: true,
                "rejectUnauthorized": false,
                headers: {
                    "Authorization": auth,
                    "Accept": "application/json",
                    "user-agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/74.0.3729.169 Safari/537.36"
                }
            };
            const result = yield request.get(options);
            yield vscode.commands.executeCommand('workbench.action.files.newUntitledFile');
            let editor = vscode.window.activeTextEditor;
            if (editor) {
                editor.edit(editBuilder => {
                    if (editor) {
                        editBuilder.insert(new vscode.Position(0, 0), JSON.parse(result).contents);
                    }
                });
            }
            vscode.window.showInformationMessage('Trace file downloaded');
        }))();
    });
}
exports.downloadTrace = downloadTrace;
function clearTrace() {
    return __awaiter(this, void 0, void 0, function* () {
        vscode.window.showInformationMessage('Clearing trace file');
        (() => __awaiter(this, void 0, void 0, function* () {
            const baseUrl = 'https://' + applianceHost + '/isam/application_logs/federation/runtime/trace.log?action=clear';
            var options = {
                method: "DELETE",
                uri: baseUrl,
                insecure: true,
                "rejectUnauthorized": false,
                headers: {
                    "Authorization": auth,
                    "Accept": "application/json",
                    "user-agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/74.0.3729.169 Safari/537.36"
                }
            };
            const result = yield request.delete(options);
            vscode.window.showInformationMessage('Trace file cleared');
        }))();
    });
}
exports.clearTrace = clearTrace;
function deployChanges(isChain) {
    return __awaiter(this, void 0, void 0, function* () {
        vscode.window.showInformationMessage('Deploying changes');
        (() => __awaiter(this, void 0, void 0, function* () {
            const baseUrl = 'https://' + applianceHost + '/isam/pending_changes';
            var options = {
                method: "PUT",
                uri: baseUrl,
                insecure: true,
                "rejectUnauthorized": false,
                headers: {
                    "Authorization": auth,
                    "Accept": "application/json",
                    "user-agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/74.0.3729.169 Safari/537.36"
                }
            };
            const result = yield request.put(options);
            vscode.window.showInformationMessage('Changes deployed');
            if (isChain) {
                yield reloadfedRuntime(true);
            }
        }))();
    });
}
exports.deployChanges = deployChanges;
function uploadMappingRule(isChain) {
    return __awaiter(this, void 0, void 0, function* () {
        if (mappingruleId !== undefined && mappingruleId !== 0) {
            vscode.window.showInformationMessage('Got mapping rule identifier, uploading...');
            let editor = vscode.window.activeTextEditor;
            let text = "";
            if (editor !== undefined) {
                text = editor.document.getText();
            }
            console.log(text);
            if (text !== undefined) {
                text = text.replace(/\"/g, '\\"');
            }
            (() => __awaiter(this, void 0, void 0, function* () {
                const baseUrl = 'https://' + applianceHost + '/iam/access/v8/mapping-rules/' + mappingruleId;
                console.log(baseUrl);
                let body_to_send = "{\"content\":\"" + text + "\"}";
                console.log(body_to_send);
                var options = {
                    body: body_to_send,
                    method: "PUT",
                    uri: baseUrl,
                    insecure: true,
                    "rejectUnauthorized": false,
                    headers: {
                        "Authorization": auth,
                        "content-type": "application/json",
                        "Accept": "application/json",
                        "user-agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/74.0.3729.169 Safari/537.36"
                    }
                };
                const result = yield request.put(options);
                vscode.window.showInformationMessage('Upload done');
                if (isChain) {
                    yield deployChanges(true);
                }
            }))();
        }
        else {
            vscode.window.showInformationMessage('No mapping rule in cache, register mapping rule');
        }
    });
}
exports.uploadMappingRule = uploadMappingRule;
function reloadfedRuntime(isChain) {
    return __awaiter(this, void 0, void 0, function* () {
        vscode.window.showInformationMessage('Reloading federation runtime');
        (() => __awaiter(this, void 0, void 0, function* () {
            const baseUrl = 'https://' + applianceHost + '/mga/runtime_profile/local/v1';
            var options = {
                body: "{operation: \"reload\"}",
                method: "PUT",
                uri: baseUrl,
                insecure: true,
                "rejectUnauthorized": false,
                headers: {
                    "Authorization": auth,
                    "Accept": "application/json",
                    "user-agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/74.0.3729.169 Safari/537.36"
                }
            };
            const result = yield request.put(options);
            vscode.window.showInformationMessage('Reloaded federation runtime');
            if (isChain) {
                vscode.window.showInformationMessage('Action chain completed');
            }
        }))();
    });
}
exports.reloadfedRuntime = reloadfedRuntime;
function getMappingRuleId() {
    return __awaiter(this, void 0, void 0, function* () {
        (() => __awaiter(this, void 0, void 0, function* () {
            const baseUrl = 'https://' + applianceHost + '/iam/access/v8/mapping-rules/?filter=name%20equals%20' + mappingruleName;
            var options = {
                method: "GET",
                uri: baseUrl,
                insecure: true,
                "rejectUnauthorized": false,
                headers: {
                    "Authorization": auth,
                    "Accept": "application/json",
                    "user-agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/74.0.3729.169 Safari/537.36"
                }
            };
            const result = yield request.get(options);
            let mappingid = JSON.parse(result)[0].id;
            if (mappingid !== undefined) {
                if (mappingid !== 0) {
                    mappingruleId = mappingid;
                    vscode.window.showInformationMessage('saved mappingrule ID');
                }
                else {
                    vscode.window.showInformationMessage('Mapping rule not found');
                    mappingruleId = 0;
                }
            }
            console.log(mappingid);
        }))();
    });
}
exports.getMappingRuleId = getMappingRuleId;
function showInputBox() {
    return __awaiter(this, void 0, void 0, function* () {
        const result = yield vscode.window.showInputBox({
            value: mappingruleName
        });
        vscode.window.showInformationMessage(`Got: ${result}`);
        if (result === undefined) {
            mappingruleName = "";
        }
        else {
            mappingruleName = result;
            getMappingRuleId();
        }
    });
}
exports.showInputBox = showInputBox;
function registerHost() {
    return __awaiter(this, void 0, void 0, function* () {
        vscode.window.showInformationMessage('Setting hostname');
        var result;
        if (applianceHost === undefined || applianceHost === "") {
            result = yield vscode.window.showInputBox({
                value: "<set your appliance hostname>"
            });
        }
        else {
            result = yield vscode.window.showInputBox({
                value: applianceHost
            });
        }
        vscode.window.showInformationMessage(`Got hostname: ${result}`);
        if (result !== undefined && result !== "") {
            applianceHost = result;
            if (username === undefined || username === "") {
                vscode.window.showInformationMessage('Setting username');
                result = yield vscode.window.showInputBox({
                    value: "admin"
                });
            }
            else {
                vscode.window.showInformationMessage('Setting username');
                result = yield vscode.window.showInputBox({
                    value: username
                });
            }
            vscode.window.showInformationMessage(`Got username: ${result}`);
            if (result !== undefined && result !== "") {
                username = result;
                if (password === undefined || password === "") {
                    vscode.window.showInformationMessage('Setting password');
                    result = yield vscode.window.showInputBox({
                        value: "<set your password>"
                    });
                }
                else {
                    result = yield vscode.window.showInputBox({
                        value: password
                    });
                }
                if (result !== undefined && result !== "") {
                    password = result;
                }
            }
        }
    });
}
exports.registerHost = registerHost;
function settingsValid() {
    if (username !== undefined && username !== "" && password !== undefined && password !== "" && applianceHost !== undefined && applianceHost !== "") {
        return true;
    }
    else {
        vscode.window.showInformationMessage('No credentials/host found, execute ISAM Register host command');
        return false;
    }
}
exports.settingsValid = settingsValid;
function activate(context) {
    let disposable = vscode.commands.registerCommand('extension.uploadMappingrule', () => {
        if (settingsValid()) {
            uploadMappingRule(false);
        }
    });
    disposable = vscode.commands.registerCommand('extension.registerMappingrule', () => {
        if (settingsValid()) {
            showInputBox();
        }
    });
    disposable = vscode.commands.registerCommand('extension.reloadRuntime', () => {
        if (settingsValid()) {
            reloadfedRuntime(false);
        }
    });
    disposable = vscode.commands.registerCommand('extension.deployChanges', () => {
        if (settingsValid()) {
            deployChanges(false);
        }
    });
    disposable = vscode.commands.registerCommand('extension.doAll', () => __awaiter(this, void 0, void 0, function* () {
        if (settingsValid()) {
            yield uploadMappingRule(true);
        }
    }));
    disposable = vscode.commands.registerCommand('extension.downloadTrace', () => {
        if (settingsValid()) {
            downloadTrace();
        }
    });
    disposable = vscode.commands.registerCommand('extension.clearTrace', () => {
        if (settingsValid()) {
            clearTrace();
        }
    });
    disposable = vscode.commands.registerCommand('extension.registerHost', () => {
        registerHost();
    });
    context.subscriptions.push(disposable);
}
exports.activate = activate;
// this method is called when your extension is deactivated
function deactivate() { }
exports.deactivate = deactivate;
//# sourceMappingURL=extension.js.map